wallet-tool: add graalvm `nativeCompile` support to `build.gradle`

msgilligan opened the pull request:
At this checkpoint the tool builds and can run well enough to display help.

To build:
```
gradle clean bitcoinj-wallettool:nativeCompile
``` 

To run:
```
./wallettool/build/native/nativeCompile/wallet-tool --help
```

TODO:
- [x] Add config/hints to get all "actions" in the tool working properly
- [x] Add GraalVM build to Github Actions

msgilligan commented:
@schildbach I think this is one of those intermittent failures, too.

msgilligan commented:
I just rebased (and squashed) on top of `msgilligan-wallettool-add-picocli-codegen` PR #2284

msgilligan commented:
PR #2284 was merged, so I rebased this PR to `master`. I also improved the detection of support for `nativeCompile` / Java Toolchain support with 07e9193505b0971829bed31895995aff02c202ff

schildbach commented on wallettool/build.gradle:
```diff
@@ -9,7 +9,10 @@ plugins {
+boolean hasAnnotationProcessor = (GradleVersion.current().compareTo(annotationProcessorMinVersion) >= 0)
+
+def toolChainsMinVersion = GradleVersion.version("6.8")     // Toolchains with selection by vendor
+boolean hasToolchains = (GradleVersion.current().compareTo(toolChainsMinVersion) >= 0)
```
Is "toolchains" a good name for this? What toolchain is it exactly? Besides, is it "toolchains" or "toolChains"?

msgilligan replied:
https://docs.gradle.org/current/userguide/toolchains.html

Toolchains were added in Gradle 6.7, but Gradle 6.8 added the JvmVendorSpec which is essential to using them in our (and most other) cases.

msgilligan replied:
So `toolChainsMinVersion` and `hasToolchains` may not be the best choice of name. The `hasToolchains` `boolean` currently "means" three things.


1. Has "Java Toolchain" support (Gradle 6.7+)
2. Has `JvmVendorSpec` support (Gradle 6.8+) in toolchains: see https://docs.gradle.org/6.8/release-notes.html#java-toolchain-improvements
3. Has "everything needed" for running the GraalVM task.

schildbach replied:
Sory, I forgot about the "toolchain" feature. I guess the name is fine then, but lets stay with the non-camelcase spelling like in your linked doc.

schildbach commented on wallettool/build.gradle:
```diff
@@ -58,9 +61,7 @@ asciidoctor {
     }
 }
 
-def minGraalVMGradleVersion = GradleVersion.version("7.3")
```
Why did this version change? Also the other version check changed too – why? A small comment in the commit description could clear this up.

msgilligan replied:
I originally used "7.3" because I was thinking to just keep things simple and use a very recent version. I later decided to switch to "6.8" because that's the actual minimum version for the feature used.

msgilligan replied:
>Also the other version check changed too – why? 

The `annotationProcessorMinVersion` check was incorrect. It was excluding "4.6" because of '> 0' so I switched it to `>= 0`.

schildbach replied:
I added a commit with the fix right away. I think fixes should stay separate, this allows them to be merged quicker and also makes this PR simpler to review.

schildbach reviewed (commented):
07e9193505b0971829bed31895995aff02c202ff looks good, with comments.

The other commit is still WIP?

msgilligan commented:
> The other commit is still WIP?

I'm not sure what you mean here. Which "other commit"?

I do still have two "TODOs" (in the first comment on this PR) pending:

> * Add config/hints to get all "actions" in the tool working properly
> * Add GraalVM build to Github Actions

msgilligan commented:
The failure in https://github.com/bitcoinj/bitcoinj/runs/5428607201?check_suite_focus=true looks unrelated to the PR. I don't remember seeing this one before, @schildbach do you know what it is?

msgilligan commented:
I'm down to one "TODO" (and addressing any of @schildbach 's comments):

> [ ] Add config/hints to get all "actions" in the tool working properly

I'm going to try using the Gradle Plugin to help generate the config: https://graalvm.github.io/native-build-tools/latest/gradle-plugin.html#agent-support

msgilligan commented:
@schildbach Shall I rebase or rebase-and-squash?

schildbach commented:
> The failure in https://github.com/bitcoinj/bitcoinj/runs/5428607201?check_suite_focus=true looks unrelated to the PR. I don't remember seeing this one before, @schildbach do you know what it is?

That's a very strange failure. The comparison fails because of something has replaced the `i` character by a `ı` chracter in (non-variable) text.

msgilligan commented:
> That's a very strange failure. The comparison fails because of something has replaced the `i` character by a `ı` chracter in (non-variable) text.

Cosmic ray? Some kind of weird CharSet issue?

msgilligan commented:
@schildbach rebased and force-pushed. I can squash if you'd like.

schildbach commented:
> > That's a very strange failure. The comparison fails because of something has replaced the `i` character by a `ı` chracter in (non-variable) text.
> 
> Cosmic ray? Some kind of weird CharSet issue?

I suspect a CharSet issue.

schildbach commented:
The PR is still marked as Draft (that's what I meant with WIP). Is this intentional?

schildbach commented:
This PR now looks clean, I'll test this real soon now.

msgilligan commented:
> The PR is still marked as Draft (that's what I meant with WIP). Is this intentional?

Yes. Because ` Add config/hints to get all "actions" in the tool working properly` is not finished yet. Not all the "actions" of the `wallet-tool` binary will work correctly until this is completed.  See https://github.com/bitcoinj/bitcoinj/pull/2285#issuecomment-1059583665

msgilligan commented:
> This PR now looks clean, I'll test this real soon now.

It would be great if you could test the build and make sure you can at least print the help message. Not all the actions work because some are reliant on reflection.

msgilligan commented:
I just noticed the non-fatal error:

```
No files were found with the provided path: /wallettool/build/native/nativeCompile/wallet-tool. No artifacts will be uploaded.
```
 in this run: https://github.com/bitcoinj/bitcoinj/actions/runs/1936458315
 
3fd7992f5167694544ce7a0dfe07d80f8290caf2 removes the leading `/` and should fix it.

msgilligan commented:
> I suspect a CharSet issue.

Oddly this is happening under macOS 11 (Big Sur), Java 17, Gradle 7.4 which is the combination I use for local development. I'm not seeing the problem on my development system.

schildbach commented:
The mechanism doesn't seem to be compatible with OpenJDK 17:
```
Execution failed for task ':bitcoinj-wallettool:nativeCompile'.
> Failed to query the value of property 'javaLauncher'.
   > No compatible toolchains found for request filter: {languageVersion=17, vendor=matching('GraalVM Community'), implementation=vendor-specific} (auto-detect true, auto-download true)
```
Using Gradle 7.0, if that matters. I just checked and GraalVM isn't in the Ubuntu repositories.

msgilligan commented:
> The mechanism doesn't seem to be compatible with OpenJDK 17:

It is compatible with OpenJDK 17, but in order to use a particular toolchain (in this case GraalVM Community Edition for language version 17) the specified JDK must either be installed locally or downloadable by Gradle (assuming downloads are enabled.) GraalVM is currently not in one of the repos that Gradle will download JDKs from.

> I just checked and GraalVM isn't in the Ubuntu repositories.

Unfortunately, this [is true](https://github.com/oracle/graal/issues/3363). So (currently) to build bitcoinj with GraalVM on Debian, GraalVM CE must be installed via another mechanism. Such as:

1. Build GraalVM from source
2. Install via the release tarball: https://www.graalvm.org/22.0/docs/getting-started/linux/
3. [SDKMan](https://sdkman.io)
4. A community-developed package such as https://github.com/dongjinleekr/graalvm-ce-deb
5. [Container images](https://github.com/graalvm/container/pkgs/container/graalvm-ce)

msgilligan commented:
@schildbach This has been rebased and updated to use the latest version of GraalVM and the latest GraalVM plugin. I've also added the `reflect-config.json` and `resource-config.json` files necessary for the `native-image` compiler to find most (hopefully) all of the needed classes. `wallet-tool` seems to work -- and is very fast -- in my preliminary testing.

I think this PR is ready to merge.

schildbach commented on wallettool/build.gradle:
```diff
@@ -48,6 +52,10 @@ if (hasJunit5) {
 }
 
+run {
+    args "--wallet=test.wallet", '--help'
```
Is there a reason you're providing those two parameters at the same time? It doesn't fail, but I think providing a wallet doesn't make sense in the `--help` case. I think the wallet will be ignored.

msgilligan replied:
To be honest, the `run` task doesn't really even need to be there. It was kind of a local modification I was using for convenience. I'll force-push an update with `run` removed entirely.

msgilligan replied:
I'm removing the entire `run` closure.

schildbach commented on wallettool/src/main/graal/resource-config.json:
```diff
@@ -0,0 +1,7 @@
+      {"pattern": ".*/english.txt$"}
+    ]
+  }
+}
```
Missing EOL here, is that intentional? If this file is generated, then the missing EOL is ok (should perhaps be reported as a bug to the generator).

msgilligan replied:
I actually used copy/paste to make this file, so it was an error on my part. I'll fix it.

schildbach commented on wallettool/src/main/graal/reflect-config.json:
```diff
@@ -0,0 +1,365 @@
+[
```
Are these files generated? If yes, from what source?

Are they even needed? Afair the previous version of your PR didn't have them and judging by the GitHub Actions reports it seemed to work.

msgilligan replied:
These files were generated by using `gradle -Pagent run` where the `run` task was configured in such a way as to run enough of the code where the agent/analyzer can discover all classes accessed via reflection, etc. I then manually copy the output into the `src/main/graal` directory.

If they are not included there will be runtime (e.g. 'class not found') errors when classes are accessed via reflection. The previous PR didn't have them, but if you actually tried to run the built tool it would fail.

msgilligan replied:
I eventually want to move these files into `bitcoinj-core` and have semi- or fully automated tools to generate them. Perhaps by using the `-Pagent` option while running unit tests. By putting them in `core` any app using **bitcoinj** will be able to produce a native-image compilation of **bitcoinj** when needed.

schildbach commented:
> 4\. A community-developed package such as https://github.com/dongjinleekr/graalvm-ce-deb

"[It's almost done.](https://github.com/oracle/graal/issues/3363#issuecomment-1083336183)"

schildbach commented:
Currently, the easiest way to review this – besides reading the source code of your contribution – seems to be looking at the results of GitHub Actions. Is this enough for now? Do we need some reviewer to actually _try_ the "graalified" wallet-tool?

msgilligan commented:
> Currently, the easiest way to review this – besides reading the source code of your contribution – seems to be looking at the results of GitHub Actions. Is this enough for now? Do we need some reviewer to actually _try_ the "graalified" wallet-tool?

I think it is probably enough for now. We should definitely consider GraalVM support as experimental/incubating and as we aren't planning on releasing a binary (at least for now) I don't think the testing/review process needs to be as rigorous as for other components. Over time, of course, that will change. There has been recent progress in providing tools for running unit tests in `native-image` mode and that is something I would like us to add in the near future.

Of course, I would love to see @schildbach and other **bitcoinj** reviewers/developers start experimenting with GraalVM and reporting and/or fixing bugs, etc. But I think the first step is to get the experimental/incubating version into `master`

msgilligan commented:
@schildbach with the latest force-push all your comments have been resolved and if you agree with my comments above (https://github.com/bitcoinj/bitcoinj/pull/2285#issuecomment-1115209385) should be ready-to-merge.
